### ISSUE TITLE
Router/PageLoader: Scroll Position is maintained on navigate when PageLoader does not "reload" content

### ISSUE THREAD

**Issue by dthyresson** [redwoodjs/redwood#1362]
As part of Issue https://github.com/redwoodjs/redwood/pull/1284 to correct the "reload flash" when navigating via Router Links, a decision needs to be made as to how to handle the window's Scroll Postition.

With change to the Router, the window position is maintained.

This means:

* Page A - Scroll 2-3 vertical screen heights
* Navigate to Page B (that also has multiple page screen height)
* You are now on Page B but at the Position of A (ie, not top)

![rw_scroll_position_issue](https://user-images.githubusercontent.com/1051633/96163708-38053900-0ee8-11eb-9886-a596bb76b21b.gif)

There is discussion with React Router w/ re: to Scroll Position and Restoration here:

https://reactrouter.com/web/guides/scroll-restoration

as well as part of browser history spec:

https://majido.github.io/scroll-restoration-proposal/history-based-api.html#web-idl

In React Router, they have a component that scrolls to top by setting the `window.scrollTo(0, 0)`:

```
class ScrollToTop extends React.Component {
  componentDidUpdate(prevProps) {
    if (
      this.props.location.pathname !== prevProps.location.pathname
    ) {
      window.scrollTo(0, 0);
    }
  }

  render() {
    return null;
  }
}
```

Perhaps this rule (or option) could be added to the `PageLoader` -- perhaps even as a Page or Route option to either a) always scroll to top or b) only scroll to top it the option is set.

**Comment by Tobbe** [comment]
Not only scroll position needs to be reset. Focus needs to be managed as well. When landing on a new page and pressing <kbd>Tab</kbd> focus should be given to the first (top-most) element on the page

**Comment by dthyresson** [comment]
My current workaround is to add to my Layout:

```
  import { useEffect } from 'react'

  ...

  useEffect(() => {
    window.scrollTo(0, 0)
  }, [])
```

**Comment by Tobbe** [comment]
This is a good write-up about things to consider
https://oaf-project.github.io/oaf-react-router/

**Comment by dthyresson** [comment]
Could/should

```
window.scrollTo(0, 0)
```

be added to Router `history.js`

```
    navigate: (to) => {
      window.history.pushState({}, null, to)
      for (const listener of Object.values(listeners)) {
        listener()
      }
    },
```

**Comment by jtoar** [comment]
Awesome discussion and great resources! Just wanted to chime in and say that I can take the lead on this since I'm doing focus management as well.

**Comment by dthyresson** [comment]
> since I'm doing focus management as well

@jtoar I looked at an older app and realized we had done something really similar to the `window.scrollTo` for focus:


```
const Page = ({ children, location }) => {
...
  const node = useRef();

  useEffect(() => {
    node.current.focus();
  }, [location.key]);
```

I cannot remember exactly what location was ... but what I do recall seeing from time to time was some some of timing :) issue when setting focus happened perhaps when the input wasn't 100% ready (this was a year or so ago for this app).

Perhaps, the timing and some defensive programming to ensure the element is ready can help.

**Comment by jtoar** [comment]
Yeah I think timing's a little tricky here, waiting for everything to be rendered before scrolling/focusing. Definitely appreciate the snippet, just to know that a solution like that's standard.

The good thing about scroll so far is that it's not a problem for popstate events (i.e. navigating backwards and forward&mdash;that scroll position's maintained), only pushstate. Still a big problem though. 

- without throttling (imperceptible!) https://s.tape.sh/ld11YgF8
- with throttling (not good at all) https://s.tape.sh/snZA8jVu

**Comment by jtoar** [comment]
@dthyresson I initially tried adding `window.scrollTo` to navigate like you said earlier but it was scrolling too soon, before things were rendered (here's a Tape https://s.tape.sh/snZA8jVu), and it was especially noticeable when you slowed the app down to fast/slow 3G.

Adding it in the `PageLoader` has all but fixed the problem so far!

**Comment by dthyresson** [comment]
> Adding it in the `PageLoader` has all but fixed the problem so far!

You are definitely right about timing -- I had seen some cases when setting focus that the element wasn't either ready or mounted (or in some way unable to be interacted with) and thus setting focus would raise an error.

Do you plan to PR this addition?
